File: DSA/DSA02029 THÁP HÀ NỘI.cpp
#include <bits/stdc++.h>
using namespace std;

void recursion(string out, string in, string other, int high)
{
    if (high <= 1)
    {
        cout << out + " -> " + in << endl;
    }
    else
    {
        recursion(out, other, in, high-1);
        cout << out + " -> " + in << endl;
        recursion(other, in, out, high-1);
    }
}

void submit()
{
    int n;
    cin >> n;

    recursion("A", "C", "B", n);
}

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    submit();

    return 0;
}

File: DSA_python/DSA02029 THÁP HÀ NỘI.py
def submit():
    n = int(input())

    def recursion(A, C, B, deep):
        if deep <= 1:
            print(A + " -> " + C);
        else:
            recursion(A, B, C, deep-1)
            print(A + " -> " + C)
            recursion(B, C, A, deep-1)

    recursion("A", "C", "B", n)




submit()